# Build with default Fedora build flags
diff --git a/scripts/build-utils.sh b/scripts/build-utils.sh
index 3a43d55..60597d0 100644
--- a/scripts/build-utils.sh
+++ b/scripts/build-utils.sh
@@ -53,7 +53,7 @@ clean_and_exit()
 
 build()
 {
-	make $OPTS CFLAGS="$CFLAGS" CPPFLAGS="$CPPFLAGS" LDFLAGS="$LDFLAGS" $@ || exitmsg "Failed build"
+	make $OPTS CFLAGS="$CFLAGS" CXXFLAGS="$CXXFLAGS" LDFLAGS="$LDFLAGS" $@ || exitmsg "Failed build"
 	return 0
 }
 
diff --git a/scripts/precompile_bitcode_files.sh b/scripts/precompile_bitcode_files.sh
index c662433..e84e123 100755
--- a/scripts/precompile_bitcode_files.sh
+++ b/scripts/precompile_bitcode_files.sh
@@ -10,22 +10,21 @@ fi
 
 # precompile instrumentation files
 FILES=
-ORIG_CPPFLAGS="$CPPFLAGS"
 for LLVM in $PREFIX/llvm-*; do
 	CLANG=$LLVM/bin/clang
 	LLVM_VERSION=${LLVM#*llvm-*}
 	INCLUDE_DIR="$LLVM/lib/clang/${LLVM_VERSION}/include/"
-	CPPFLAGS="-I ${INCLUDE_DIR} $ORIG_CPPFLAGS"
+	CPPFLAGS="-I ${INCLUDE_DIR}"
 	for F in `find $INSTR/instrumentations/ -name '*.c'`; do
 		NAME=`basename $F`
 		OUT=${NAME%*.c}.bc
 		mkdir -p "$LLVM/lib" "$LLVM/lib32"
 
 		FILES="$FILES ${LLVM#install/}/lib/$OUT"
-		$CLANG $CPPFLAGS -O3 -emit-llvm -c $F -o $LLVM/lib/$OUT $CPPFLAGS $CFLAGS $LDFLAGS
+		$CLANG $CPPFLAGS -O3 -emit-llvm -c $F -o $LLVM/lib/$OUT
 
 		FILES="$FILES ${LLVM#install/}/lib32/$OUT"
-		$CLANG $CPPFLAGS -O3 -emit-llvm -c $F -m32 -o $LLVM/lib32/$OUT $CPPFLAGS $CFLAGS $LDFLAGS
+		$CLANG $CPPFLAGS -O3 -emit-llvm -c $F -m32 -o $LLVM/lib32/$OUT
 	done
 done
 
@@ -34,18 +33,18 @@ for LLVM in $PREFIX/llvm-*; do
 	CLANG=$LLVM/bin/clang
 	LLVM_VERSION=${LLVM#*llvm-*}
 	INCLUDE_DIR="$LLVM/lib/clang/${LLVM_VERSION}/include/"
-	CPPFLAGS="-I ${INCLUDE_DIR} -Iinclude/ $ORIG_CPPFLAGS"
+	CPPFLAGS="-I ${INCLUDE_DIR} -Iinclude/"
 	for F in `find $LIBS -name '*.c'`; do
 		NAME=`basename $F`
 		OUT="${F#*/}" # strip the lib/ prefix
 		OUT="${OUT%*.c}.bc" # change .c for .bc
 
 		mkdir -p "$(dirname $LLVM/lib/$OUT)"
-		$CLANG $CPPFLAGS -O3 -emit-llvm -c $F -o $LLVM/lib/$OUT $CPPFLAGS $CFLAGS $LDFLAGS
+		$CLANG $CPPFLAGS -O3 -emit-llvm -c $F -o $LLVM/lib/$OUT
 		FILES="$FILES ${LLVM#install/}/lib/$OUT"
 
 		mkdir -p "$(dirname $LLVM/lib32/$OUT)"
-		$CLANG $CPPFLAGS -O3 -emit-llvm -c $F -m32 -o $LLVM/lib32/$OUT $CPPFLAGS $CFLAGS $LDFLAGS
+		$CLANG $CPPFLAGS -O3 -emit-llvm -c $F -m32 -o $LLVM/lib32/$OUT
 		FILES="$FILES ${LLVM#install/}/lib32/$OUT"
 	done
 done

# Fix build with GCC 11
diff --git a/dg/include/dg/PointerAnalysis/Pointer.h b/dg/include/dg/PointerAnalysis/Pointer.h
index 2a05d3ac..1cc1e862 100644
--- a/dg/include/dg/PointerAnalysis/Pointer.h
+++ b/dg/include/dg/PointerAnalysis/Pointer.h
@@ -3,6 +3,7 @@
 
 #include "dg/Offset.h"
 #include <cassert>
+#include <cstddef>
 
 namespace dg {
 namespace pta {
