--- a/scripts/symbiotic
+++ b/scripts/symbiotic
@@ -23,7 +23,7 @@ import os
 from time import time

 # set path to our package
-pth = os.path.join(os.path.dirname(__file__), '../lib/symbioticpy')
+pth = os.path.join(os.path.dirname(os.readlink(__file__)), '../lib/symbioticpy')
 sys.path.append(os.path.abspath(pth))

 from symbiotic.utils import err, dbg

# Find coorrect LLVM cmaka dir on rpm-based distros
diff --git a/system-build.sh b/system-build.sh
index aee9cc2..27ad938 100755
--- a/system-build.sh
+++ b/system-build.sh
@@ -234,12 +234,7 @@ export LLVM_PREFIX="$PREFIX/llvm-$LLVM_VERSION"
 LLVM_MAJOR_VERSION="${LLVM_VERSION%%\.*}"
 LLVM_MINOR_VERSION=${LLVM_VERSION#*\.}
 LLVM_MINOR_VERSION="${LLVM_MINOR_VERSION%\.*}"
-LLVM_CMAKE_CONFIG_DIR=share/llvm/cmake
-if [ $LLVM_MAJOR_VERSION -gt 3 ]; then
-	LLVM_CMAKE_CONFIG_DIR=lib/cmake/llvm
-elif [ $LLVM_MAJOR_VERSION -ge 3 -a $LLVM_MINOR_VERSION -ge 9 ]; then
-	LLVM_CMAKE_CONFIG_DIR=lib/cmake/llvm
-fi
+LLVM_CMAKE_CONFIG_DIR="$($LLVM_CONFIG --cmakedir)"
 
 mkdir -p $LLVM_PREFIX/bin
 for T in $LLVM_TOOLS; do
@@ -351,7 +346,7 @@ if [ $FROM -le 6 -a "$BUILD_PREDATOR" = "yes" ]; then
 	pushd predator-${LLVM_VERSION}
 
 	if [ ! -d CMakeFiles ]; then
-	        CXX=clang++ ./switch-host-llvm.sh /usr/${LLVM_CMAKE_CONFIG_DIR}
+	        CXX=clang++ ./switch-host-llvm.sh "${LLVM_CMAKE_CONFIG_DIR}"
 	fi
 
        	build || exit 1
