# sbt-instrumentation
diff --git a/sbt-instrumentation/src/instr.cpp b/sbt-instrumentation/src/instr.cpp
index 282637d..d32be1d 100644
--- a/sbt-instrumentation/src/instr.cpp
+++ b/sbt-instrumentation/src/instr.cpp
@@ -309,9 +309,8 @@ bool cloneMetadata(const llvm::Instruction *i1, llvm::Instruction *i2) {
         return cloneMetadata(pred->getTerminator(), i2);
     } else {
       DebugLoc DL;
-      if (auto SP = i1->getParent()->getParent()->getSubprogram()) {
-        DL = DebugLoc::get(SP->getScopeLine(), 0, SP);
-      }
+      if (auto SP = i1->getParent()->getParent()->getSubprogram())
+        DL = DILocation::get(SP->getContext(), SP->getScopeLine(), 0, SP);
       i2->setDebugLoc(DL);
     }
 
