# sbt-transforms
diff --git a/transforms/CMakeLists.txt b/transforms/CMakeLists.txt
index 4f71902..2e267f7 100644
--- a/transforms/CMakeLists.txt
+++ b/transforms/CMakeLists.txt
@@ -93,6 +93,7 @@
 
 # remove lib prefix for compatibility with older releases
 set_target_properties(LLVMsbt PROPERTIES PREFIX "")
+target_link_libraries(LLVMsbt LLVM)
 
 install(TARGETS LLVMsbt
         LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR})
