# sbt-transforms
diff --git a/transforms/CMakeLists.txt b/transforms/CMakeLists.txt
index 4f71902..2e267f7 100644
--- a/transforms/CMakeLists.txt
+++ b/transforms/CMakeLists.txt
@@ -9,14 +9,6 @@ message(STATUS "Using LLVMConfig.cmake in: ${LLVM_DIR}")
 list(APPEND CMAKE_MODULE_PATH "${LLVM_DIR}")
 list(APPEND CMAKE_MODULE_PATH "${LLVM_CMAKE_DIR}")
 
-# This is probably unnecessary
-if (${LLVM_PACKAGE_VERSION} VERSION_LESS "3.8")
-  # needed for LLVM 3.7
-  set(LLVM_RUNTIME_OUTPUT_INTDIR ${CMAKE_BINARY_DIR}/${CMAKE_CFG_INTDIR})
-  set(LLVM_LIBRARY_OUTPUT_INTDIR ${CMAKE_BINARY_DIR}/${CMAKE_CFG_INTDIR})
-  include(HandleLLVMOptions)
-endif()
-
 include(AddLLVM)
 
 # explicitly add -std=c++11 (c++14), some cmake versions
@@ -57,11 +50,6 @@ include(GNUInstallDirs)
 message(STATUS "CMAKE_INSTALL_LIBDIR: \"${CMAKE_INSTALL_LIBDIR}\"")
 message(STATUS "CMAKE_INSTALL_INCLUDEDIR: \"${CMAKE_INSTALL_INCLUDEDIR}\"")
 
-# We need all the symbols with dynamic libs. With static libs, we get errors.
-if (BUILD_SHARED_LIBS)
-  set(LLVM_LINK_COMPONENTS analysis core support)
-endif()
-
 add_library(LLVMsbt MODULE "BreakCritLoops.cpp"
                            "BreakInfiniteLoops.cpp"
                            "AInliner.cpp"
@@ -92,6 +80,7 @@ add_library(LLVMsbt MODULE "BreakCritLoops.cpp"
 
 # remove lib prefix for compatibility with older releases
 set_target_properties(LLVMsbt PROPERTIES PREFIX "")
+target_link_libraries(LLVMsbt LLVM)
 
 install(TARGETS LLVMsbt
         LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR})
